### DISCUSSION TITLE
Is it possible to use the function of decrypting encrypted logs as a plugin?

### DISCUSSION THREAD

**Discussion by George-J-Cho** [COVESA/dlt-viewer#358] [Q&A]
I want to decrypt the encrypted log and output it to Payload.

Is it possible to make the decryption feature a plugin?
If so, please let me know how.

For example, I want to output the payload string by performing shifting, base64, or AES.

I would be grateful if you could give me any advice on the subject matter.

Please help me resolve this issue

**Comment by alexmucde** [comment]
@George-J-Cho If your DLT File is fully encrypted, this is not possible with a plugin. If only the payload is encrypted, this is possible with a decoder plugin. You have to decrypt the payload and then parse the arguments.

**Reply by George-J-Cho** [reply]
Thank you very much for your reply.
Can you give me some simple advice on how to use the decode plugin?
Or can you advise which documentation I should refer to?
I am currently reading the document "dlt_viewer_plugins_programming_guide.pdf"

**Comment by George-J-Cho** [comment]
I wrote the code below and built it, using the
I turned on the plugin as shown in the screenshot below
The plugin didn't do what I wanted. 

Someone give me some advice

<hr>

dummydecoderplugin.cpp
```
bool DummyDecoderPlugin::decodeMsg(QDltMsg &msg, int triggeredByUser)
{
    QString str_ctid;
    str_ctid = msg.getCtid();
    
    // "String" is the value of the ctid
    if (str_ctid == "String" ) 
    {
        QByteArray encodedPayload = msg.getPayload();

        QByteArray shiftedPayload;
        for (int i = 0; i < encodedPayload.size(); i++)
        {
            shiftedPayload.append(encodedPayload.at(i) - 3);
        }

        QByteArray decodedPayload = QByteArray::fromBase64(shiftedPayload);

        msg.setPayload(decodedPayload);
        return true;
    }
    Q_UNUSED(msg);
    Q_UNUSED(triggeredByUser);

    return false;
}
```


![image](https://github.com/COVESA/dlt-viewer/assets/109511602/ae680750-5f6e-4b40-a079-be42e438beba)

**Comment by alexmucde** [comment]
Hi George,

i think something like this is missing:


        /* get the arguments of the payload */
        if(mode==DltModeVerbose) {
            offset = 0;
            arguments.clear();
            for(int num=0;num<numberOfArguments;num++) {
                if(argument.setArgument(payload,offset,endianness)==false) {
                    /* There was an error parsing the arguments */
                    return false;
                }
                arguments.append(argument);
            }
        }

But  i think you have no access to arguments. The class QDltMsg must be extended to get access.

**Reply by alexmucde** [reply]
Perhaps we can create an new function of QDltMsg which updates the arguments.

**Comment by George-J-Cho** [comment]
@alexmucde 
thank you.
I worked on the desired functionality in "tablemodel.cpp" on behalf of the plugin.

What you explained above about plugins is something I'll have to think about.
Thanks again for your response.

**Comment by George-J-Cho** [comment]
In the code below, the string shown in the "DLT-Viewer" appears broken
How should I write it so it doesn't break?

By default, incoming data is base64 encoded.
If you copy and paste the broken content into Notepad or the Web, you'll see the string as normal

I googled for help and haven't solved it in a week.
Please help me...

tablemodel.cpp
```                
/* display payload */
QString visu_data; 
visu_data = msg.toStringPayload().trimmed().replace('\n', ' ');

QByteArray encodedData = visu_data.toUtf8();
QByteArray lastFive = encodedData.right(5);
encodedData.remove(encodedData.length() - 5, 5);
encodedData.prepend(lastFive);
// visu_data = encodedData;
// If you put a value here, like the contents of a comment, it won't break the string.

QByteArray decodedBytes = QByteArray::fromBase64(encodedData);
visu_data = decodedBytes;

```
